promql/engine.go

```diff
@@ -697,117 +655,355 @@ func (ev *evaluator) eval(expr Expr) Value {
-		return ev.aggregation(e.Op, e.Grouping, e.Without, e.Param, Vector)
+		if s, ok := e.Param.(*StringLiteral); ok {
+			return rangeWrapper(func(v []Value, enh *evalNodeHelper) Vector {
+				return ev.aggregation(e.Op, e.Grouping, e.Without, s.Val, v[0].(Vector), enh)
+			}, e.Expr)
+		}
+		return rangeWrapper(func(v []Value, enh *evalNodeHelper) Vector {
+			var param float64
+			if e.Param != nil {
+				param = v[0].(Vector)[0].V
+			}
+			return ev.aggregation(e.Op, e.Grouping, e.Without, param, v[1].(Vector), enh)
+		}, e.Param, e.Expr)
 
-	case *BinaryExpr:
-		lhs := ev.evalOneOf(e.LHS, ValueTypeScalar, ValueTypeVector)
-		rhs := ev.evalOneOf(e.RHS, ValueTypeScalar, ValueTypeVector)
+	case *Call:
+		if e.Func.Name == "timestamp" {
+			// Matrix evaluation always returns the evaluation time,
+			// so this function needs special handling when given
+			// a vector selector.
+			vs, ok := e.Args[0].(*VectorSelector)
+			if ok {
+				return rangeWrapper(func(v []Value, enh *evalNodeHelper) Vector {
+					return e.Func.Call([]Value{ev.vectorSelector(vs, enh.ts)}, e.Args, enh)
+				})
+			}
+		}
+		// Check if the function has a matrix argument.
+		for matrixArgIndex, a := range e.Args {
+			sel, ok := a.(*MatrixSelector)
+			if ok {
+				inArgs := make([]Value, len(e.Args))
+				// Evaluate any non-matrix arguments.
+				otherArgs := make([]Matrix, len(e.Args))
+				otherArgsIn := make([]Vector, len(e.Args))
+				for i, e := range e.Args {
+					if i != matrixArgIndex {
+						otherArgs[i] = ev.eval(e).(Matrix)
+						otherArgsIn[i] = Vector{Sample{}}
+						inArgs[i] = otherArgsIn[i]
+					}
+				}
 
-		switch lt, rt := lhs.Type(), rhs.Type(); {
-		case lt == ValueTypeScalar && rt == ValueTypeScalar:
-			return Scalar{
-				V: scalarBinop(e.Op, lhs.(Scalar).V, rhs.(Scalar).V),
-				T: ev.Timestamp,
+				mat := make(Matrix, 0, len(sel.series)) // Output matrix.
+				offset := durationMilliseconds(sel.Offset)
+				// Reuse objects across steps to save memory allocations.
+				points := getPointSlice(16)
+				inMatrix := make(Matrix, 1)
+				inArgs[matrixArgIndex] = inMatrix
+				enh := &evalNodeHelper{out: make(Vector, 0, 1)}
+				// Process all the calls for one time series at a time.
+				var it *storage.BufferedSeriesIterator
+				for i, s := range sel.series {
+					if it == nil {
+						it = storage.NewBuffer(s.Iterator(), durationMilliseconds(sel.Range))
+					} else {
+						it.Reset(s.Iterator())
+					}
+					ss := Series{
+						// For all range vector functions, the only change to the
+						// output labels is dropping the metric name so just do
+						// it once here.
+						Metric: dropMetricName(sel.series[i].Labels()),
+						Points: getPointSlice(numSteps),
+					}
+					inMatrix[0].Metric = sel.series[i].Labels()
+					for ts, step := ev.Timestamp, -1; ts <= ev.EndTimestamp; ts += ev.Interval {
+						step++
+						// Set the non-matrix arguments.
+						// They are scalar, so it is safe to use the step number
```

Comment:
If they're all scalar, why bother using the step number at all? They are all going to be equal anyway. You might as well set the value only once, outside the for loops (line 792 above).

Reply:
Oops, my bad. I had briefly considered the possibility, but dismissed it before I actually thought it through, on the grounds that you mention.

Reply:
It's a bit annoying code wise, but those are the documented semantics. I don't think it's something we can realistically remove, as someone might want to do `log(2)` as a value or something which is reasonable.